internal/pruner/pruner.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package pruner
+
+import (
+	"path/filepath"
+	"strings"
+
+	"github.com/syncthing/protocol"
+	"github.com/syncthing/syncthing/internal/db"
+)
+
+type Pruner struct {
+	dirPatterns  []string
+	filePatterns []string
+}
+
+func New(patterns []string) *Pruner {
+	p := &Pruner{
+		dirPatterns:  make([]string, 0),
+		filePatterns: make([]string, 0),
+	}
+	for _, pattern := range patterns {
+		if strings.HasSuffix(pattern, "[^/]+") {
+			p.filePatterns = append(p.filePatterns, filepath.Dir(pattern))
+		} else {
+			p.dirPatterns = append(p.dirPatterns, pattern)
+		}
+	}
+	return p
+}
+
+func (p *Pruner) ShouldSkipFile(path string) bool {
+	for _, pattern := range p.dirPatterns {
+		// A file which is within a given directory pattern
+		if strings.HasPrefix(path, pattern) {
+			return false
+		}
+	}
+
+	dir := filepath.Dir(path)
+
+	for _, pattern := range p.filePatterns {
+		// A file which matches one of the file patterns
+		if pattern == dir {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *Pruner) ShouldSkipDirectory(path string) bool {
```

Comment:
Also, I guess I should add `/` to all dir patterns and to all dir paths passed in.